llvm/lib/IR/Instruction.cpp

```diff
@@ -386,11 +394,17 @@ void Instruction::setNonNeg(bool b) {
 }
 
 bool Instruction::hasNoUnsignedWrap() const {
-  return cast<OverflowingBinaryOperator>(this)->hasNoUnsignedWrap();
+  if (auto Inst = cast<OverflowingBinaryOperator>(this))
+    return Inst->hasNoUnsignedWrap();
+
+  return cast<PossiblyNoWrapInst>(this)->hasNoUnsignedWrap();
 }
 
 bool Instruction::hasNoSignedWrap() const {
-  return cast<OverflowingBinaryOperator>(this)->hasNoSignedWrap();
+  if (auto Inst = cast<OverflowingBinaryOperator>(this))
+    return Inst->hasNoSignedWrap();
+
+  return cast<PossiblyNoWrapInst>(this)->hasNoSignedWrap();
```

Comment:
Can there be a common base class for has nuw/nsw instead of requiring this split?

Reply:
We can change the name of `OverflowingBinaryOperator` class to something 
that works for `trunc` too, but I am afraid that they will be handled differently later in the pipeline.

Reply:
I don't think it makes sense to share them. Outside of the flag setting/getting logic here, their use will be pretty much completely disjoint.